capsules/src/virtual_hmac.rs

```diff
@@ -0,0 +1,104 @@
+//! Virtualize the HMAC interface to enable multiple users of an underlying
+//! HMAC hardware peripheral.
+
+use kernel::ReturnCode;
+use core::cell::Cell;
+use kernel::common::cells::OptionalCell;
+use kernel::common::{List, ListLink, ListNode};
+use kernel::hil::hmac;
+
+pub struct VirtualMuxHmac<'a, A: hmac::HMAC<'a>> {
+    mux: &'a MuxHmac<'a, A>,
+    key: OptionalCell<&'static [u8]>,
+    data: OptionalCell<&'static [u8]>,
+    data_len: OptionalCell<usize>,
+    next: ListLink<'a, VirtualMuxHmac<'a, A>>,
+    client: OptionalCell<&'a dyn hmac::Client<'a>>,
+}
+
+impl<A: hmac::HMAC<'a>> ListNode<'a, VirtualMuxHmac<'a, A>> for VirtualMuxHmac<'a, A> {
+    fn next(&self) -> &'a ListLink<VirtualMuxHmac<'a, A>> {
+        &self.next
+    }
+}
+
+impl<A: hmac::HMAC<'a>> VirtualMuxHmac<'a, A> {
+    pub fn new(mux_alarm: &'a MuxHmac<'a, A>) -> VirtualMuxHmac<'a, A> {
+        VirtualMuxHmac {
+            mux: mux_alarm,
+            key: OptionalCell::empty(),
+            data: OptionalCell::empty(),
+            data_len: OptionalCell::empty(),
+            next: ListLink::empty(),
+            client: OptionalCell::empty(),
+        }
+    }
+}
+
+impl<A: hmac::HMAC<'a>> hmac::HMAC<'a> for VirtualMuxHmac<'a, A> {
+    fn set_client(&'a self, client: &'a dyn hmac::Client<'a>) {
+        self.client.set(client);
+    }
+
+    /// Set the encryption key.
+    /// Returns `EINVAL` if length is not `HMAC_KEY_SIZE`
+    fn set_key(&self, key: &'static [u8]) -> ReturnCode {
+        self.key.set(key);
+
+        ReturnCode::SUCCESS
+    }
+
+    /// Add data to the hmac IP.
+    /// All data passed in is fed to the HMAC hardware block.
+    /// Returns the number of bytes written on success
+    fn add_data(&self, data: &'static mut [u8], length: usize) -> Result<usize, ReturnCode> {
+        self.data.set(data);
+        self.data_len.set(length);
+
+        Ok(0)
```

Comment:
The comment says the number of bytes written, is that supposed to be 0 here?

Reply:
It shouldn't be, thanks